source/AI.cpp

```diff
@@ -309,6 +312,199 @@ AI::AI(const List<Ship> &ships, const List<Minable> &minables, const List<Flotsa
 
 
 // Fleet commands from the player.
+void AI::IssueFormationChange(const PlayerInfo &player)
+{
+	// Figure out what ships we are giving orders to
+	vector<Ship *> ships;
+	bool fullFleet = player.SelectedShips().empty();
+	if(fullFleet)
+	{
+		for(const shared_ptr<Ship> &it : player.Ships())
+			if(it.get() != player.Flagship() && !it->IsParked()
```

Comment:
Formations are also applied to ships that are carried (and they have by default their carrier as parent, not the players flagship).

The "Get currently selected ships or whole fleet" functions are present in this code multiple times, but each selection seems to be a little bit different;
- the loop for `cloaking` has an additional check on "can cloak"
- the `AI::IssueOrders` loop doesn't seem to check for destroyed (to include destroyed ships for clearing orders)
- the `IssueDeploy` loop has an additional check to only select carried ships.

The only 3 functions that have the same selection criteria are the IssueFormation* commands ('PatternSelect', 'RingIncrease' and `RingDecrease`). I could write a common function for those 3 or I could try to create a more generic function that also gets a lambda function parameter to allow additional criteria to be set on such generic function.
If we go for the generic function, then I might want to keep that for a separate PR (because it has the potential to increase the likelyhood of merge conflicts).

Reply:
Ok makes sense, thanks for clarifying.

Reply:
I will add the common function for the 3 formation functions with name `GetShipsForFormationCommand` (in the anonymous namespace of AI).